python/paddle/fluid/tests/unittests/test_program_prune_backward.py

```diff
@@ -0,0 +1,190 @@
+    fc0 = fluid.layers.fc(input=emb, size=hid_dim * 4)
+    lstm_h, c = fluid.layers.dynamic_lstm(
+        input=fc0, size=hid_dim * 4, is_reverse=False)
+    lstm_max = fluid.layers.sequence_pool(input=lstm_h, pool_type='max')
+    lstm_max_tanh = fluid.layers.tanh(lstm_max)
+    fc1 = fluid.layers.fc(input=lstm_max_tanh, size=hid_dim2, act='tanh')
+    prediction = fluid.layers.fc(input=fc1, size=class_dim, act='softmax')
+    cost = fluid.layers.cross_entropy(input=prediction, label=label)
+    avg_cost = fluid.layers.mean(x=cost)
+    return avg_cost
+
+
+class TestProgramPruneBackward(unittest.TestCase):
+    def program_compare(self, program_a, program_b):
+        assert isinstance(
+            program_a, fluid.framework.
+            Program), "The first argument should be fluid.framework.Program."
+        assert isinstance(
+            program_b, fluid.framework.
+            Program), "The second argument should be fluid.framework Program."
+
+        self.assertEqual(len(program_a.blocks), len(program_b.blocks))
+        for idx in range(len(program_a.blocks)):
+            block_a = program_a.blocks[idx]
+            block_b = program_b.blocks[idx]
+            self.assertEqual(len(block_a.ops), len(block_b.ops))
+            for op_idx in range(len(block_a.ops)):
+                self.assertEqual(block_a.ops[op_idx].type,
+                                 block_b.ops[op_idx].type)
+            for var_key in list(block_a.vars.keys()):
+                # FIXME(chenweihang): these variables are independent in scope, and used for cache.
+                # Because they are not associated with op, the prune function can't find them.
+                # Pruning these variabels doesn't affect the correctness of execution
+                if block_a.vars[var_key].name == "kCUDNNFwdAlgoCache" or \
+                        block_a.vars[var_key].name == "kCUDNNBwdDataAlgoCache" or \
+                        block_a.vars[var_key].name == "kCUDNNBwdFilterAlgoCache":
+                    continue
+                self.assertTrue(block_b.has_var(var_key))
+
+    def check_prune_correctness(self, method, feed_dict, optimizer):
+        loss = method(use_feed=False)
+
+        main_program = fluid.default_main_program()
+        test_prog_orig = main_program.clone(for_test=True)
+        optimizer().minimize(loss)
+        test_prog_prune = main_program.clone(for_test=True)
+        self.program_compare(test_prog_orig, test_prog_prune)
+
+        place = core.CPUPlace()
+        exe = fluid.Executor(place)
+        exe.run(fluid.default_startup_program())
+
+        loss_data_orig, = exe.run(test_prog_orig,
+                                  feed=feed_dict,
+                                  fetch_list=[loss.name])
+        loss_data_prune, = exe.run(test_prog_prune,
+                                   feed=feed_dict,
+                                   fetch_list=[loss.name])
+        self.assertEqual(loss_data_orig, loss_data_prune)
```

Comment:
You should run `test_prog_prune` first here.
Because, if the pruning function has a bug, which induces to the `test_prog_prune` has backward ops and optimization ops, the parameters of this model may be changed.

Reply:
Done